lib/node-token-validation/src/api/TokenValidator.ts

```diff
@@ -0,0 +1,170 @@
+    private config: TokenValidationConfiguration;
+    protected logger: Logger;
+    protected networkInterface: INetworkModule;
+    protected openIdConfigProvider: OpenIdConfigProvider;
+
+    constructor(configuration: Configuration) {
+        this.config = buildConfiguration(configuration);
+        this.logger = new Logger(this.config.system.loggerOptions, name, version);
+        this.networkInterface = this.config.system.networkClient,
+        this.openIdConfigProvider = new OpenIdConfigProvider(this.config, this.networkInterface, this.logger);
+    }
+
+    async validateToken(token: string, options: TokenValidationParameters): Promise<TokenValidationResponse> {
+        this.logger.trace("TokenValidator.validateToken called");
+        
+        if (!token || token.length < 1) {
+            throw ValidationConfigurationError.createMissingTokenError();
+        }
+
+        const validationParams: ValidationParameters = await buildTokenValidationParameters(options);
+        this.logger.verbose("TokenValidator - ValidationParams built");
+            
+        const jwks = await this.getJWKS(validationParams);
+
+        const jwtVerifyParams: JWTVerifyOptions = {
+            algorithms: validationParams.validAlgorithms,
+            issuer: await this.setIssuerParams(options),
+            audience: await this.setAudienceParams(options),
+            subject: validationParams.subject,
+            typ: validationParams.validTypes[0]
+        };
+
+        const { payload, protectedHeader } = await jwtVerify(token, jwks, jwtVerifyParams);
+
+        this.validateClaims(payload, validationParams);
+
+        return {
+            protectedHeader,
+            payload,
+            token,
+            tokenType: validationParams.validTypes[0]
+        };
+    }
+    
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    async getJWKS(validationParams: ValidationParameters): Promise<any> {
+        this.logger.trace("TokenValidator.getJWKS called");
+        
+        // Prioritize keystore or jwksUri if provided
+        if (validationParams.issuerSigningKeys) {
+            this.logger.verbose("TokenValidator - issuerSigningKeys provided");
+            return createLocalJWKSet({
+                keys: validationParams.issuerSigningKeys
+            });
+        } 
+        
+        if (validationParams.issuerSigningJwksUri) {
+            this.logger.verbose("TokenValidator - Creating JWKS from JwksUri");
+            return createRemoteJWKSet(new URL(validationParams.issuerSigningJwksUri));
+        }
+
+        // Do resiliency well-known endpoint thing here
+        const retrievedJwksUri: string = await this.openIdConfigProvider.fetchJwksUriFromEndpoint();
+        this.logger.verbose("TokenValidator - Creating JWKS from default");
+        return createRemoteJWKSet(new URL(retrievedJwksUri));
+    }
+
+    async setIssuerParams(options: TokenValidationParameters): Promise<string[]> {
```

Comment:
Why is this async? Is this something that will be added onto later?

Reply:
+1, not sure why these functions are async or return promises

Reply:
Ah, this was left over from when we were going to use custom validators. Changed these back to sync. Thanks for the catch.